### PR TITLE
feat: CSStrikes

### PR DESCRIPTION
**Pull request by vgorkavenko**

### CODE REVIEW

File: src/CSModule.sol [hunk 1]
```diff
@@ -109,14 +114,17 @@ contract CSModule is
...
         address admin
     ) external initializer {
         if (_accounting == address(0)) revert ZeroAccountingAddress();
+        if (_strikes == address(0)) revert ZeroStrikesAddress();
         if (admin == address(0)) revert ZeroAdminAddress();
 
         __AccessControlEnumerable_init();
 
         accounting = ICSAccounting(_accounting);
+        strikes = ICSStrikes(_strikes);
```
Comment (dgusakov) [inline src/CSModule.sol:127]: A separate initialise_v2 will be required

File: src/CSModule.sol [hunk 2]
```diff
@@ -1128,7 +1187,7 @@ contract CSModule is
         uint256 nodeOperatorId,
         uint256 startIndex,
         uint256 keysCount
-    ) external view returns (bytes memory) {
+    ) public view returns (bytes memory) {
```
Comment (dgusakov) [inline src/CSModule.sol:1190]: Fix ordering warn below

File: src/CSModule.sol [hunk 3]
```diff
@@ -1388,6 +1447,22 @@ contract CSModule is
...
+    ) internal view returns (bool) {
+        uint256 strikesDataLength = strikesData.length;
+        if (strikesDataLength != strikesLifetime) revert InvalidStrikesData();
+        uint256 count = 0;
+        for (uint256 i = 0; i < strikesDataLength; i++) {
+            if (strikesData[i] > 0) {
+                count++;
+            }
+        }
+        return count > ejectThreshold - 1;
```
Comment (dgusakov) [inline src/CSModule.sol:1463]: AFAIK this is cheaper
```suggestion
        return count >= ejectThreshold;
```

File: src/CSModule.sol (hunk 1 above)
Comment (dgusakov) [reply src/CSModule.sol:127]: It will be better to add a setter since you have one in the Oracle

File: src/CSFeeOracle.sol [hunk 4]
```diff
@@ -65,6 +70,13 @@ contract CSFeeOracle is
         _setFeeDistributorContract(feeDistributorContract);
     }
 
+    /// @inheritdoc ICSFeeOracle
+    function setStrikesContract(
```
Comment (skhomuti) [inline src/CSFeeOracle.sol:74]: why do we have the ability to change it here but not in the module

File: src/CSModule.sol [hunk 5]
```diff
@@ -1388,6 +1447,22 @@ contract CSModule is
         }
     }
 
+    function _isEnoughStrikesToEject(
+        uint256 strikesLifetime,
+        uint256 ejectThreshold,
+        uint256[] calldata strikesData
+    ) internal view returns (bool) {
+        uint256 strikesDataLength = strikesData.length;
+        if (strikesDataLength != strikesLifetime) revert InvalidStrikesData();
```
Comment (skhomuti) [inline src/CSModule.sol:1456]: this check is not related, put it on the level above

File: src/CSStrikes.sol [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        bool isSameCid = keccak256(bytes(_treeCid)) ==
+            keccak256(bytes(treeCid));
+        bool isSameRoot = _treeRoot == treeRoot;
+
+        if (bytes(_treeCid).length != 0 && _treeRoot != bytes32(0)) {
+            if (isSameCid) revert InvalidTreeCID();
+            if (isSameRoot) revert InvalidTreeRoot();
+        }
+
+        if (!isSameCid && !isSameRoot) {
```
Comment (skhomuti) [inline src/CSStrikes.sol:60]: that's a pretty strange condition, don't get it
it allows to ignore setting if any of these are the same, but it seems to be an illegal values, isn't it? in which cases we will receive such values?

File: src/interfaces/ICSModule.sol [hunk 7]
```diff
@@ -479,6 +488,20 @@ interface ICSModule is IQueueLib, INOAddresses, IAssetRecovererLib {
         bool isSlashed
     ) external;
 
+    /// @notice Report Node Operator's key as withdrawn and settle withdrawn amount
+    /// @notice Called by the CSStrikes contract.
+    ///         See `CSStrikes.processValidatorEjection` to use this method permissionless
+    /// @param nodeOperatorId ID of the Node Operator
+    /// @param keyIndex Index of the withdrawn key in the Node Operator's keys storage
+    /// @param strikesData Strikes of the Node Operator
```
Comment (skhomuti) [inline src/interfaces/ICSModule.sol:496]: it's about node operator keys i believe
it's not clear what are the values supposed to be in this list

File: src/CSModule.sol [hunk 8]
```diff
@@ -862,6 +870,49 @@ contract CSModule is
...
+            revert AlreadySubmitted();
+        }
+
+        (uint256 lifetime, uint256 threshold) = PARAMETERS_REGISTRY
+            .getStrikesParams(accounting.getBondCurveId(nodeOperatorId));
+        // TODO: get from somewhere
+        uint256 penaltyAmount = 0;
+        //
+
+        bool isEnoughStrikes = _isEnoughStrikesToEject(
```
Comment (skhomuti) [inline src/CSModule.sol:897]: bad naming here

File: src/CSModule.sol (hunk 5 above)
Comment (skhomuti) [inline src/CSModule.sol:1456]: what stops me of sending a proof for a stale key, such as:
- these are not the last X strikes but some from the past frames
- key has been withdrawn or exited already

**Review by skhomuti — CHANGES_REQUESTED** [review]

File: src/CSModule.sol (hunk 5 above)
Comment (vgorkavenko) [reply src/CSModule.sol:1456]: Oracle is in charge of setting the root with actual and valid strikes

File: src/CSStrikes.sol (hunk 6 above)
Comment (vgorkavenko) [reply src/CSStrikes.sol:60]: We don't want to update values from 0 to 0. But in the same time the new one can be empty  when the old one not

File: src/CSModule.sol (hunk 2 above)
Comment (vgorkavenko) [reply src/CSModule.sol:1190]: typo. will be reverter

File: src/CSStrikes.sol (hunk 6 above)
Comment (vgorkavenko) [reply src/CSStrikes.sol:60]: > If you have the same root, you should have the same CID, right?

Yes

File: src/CSModule.sol (hunk 5 above)
Comment (vgorkavenko) [reply src/CSModule.sol:1456]: > This check is off and redundant in general. Oracle should take care of expired strikes and remove them from the tree

We should care about case when `lifetime` was changed in the contract, but new report is not happened yet

File: src/CSStrikes.sol (hunk 6 above)
Comment (skhomuti) [reply src/CSStrikes.sol:60]: so why do we have this condition here then

Comment (vgorkavenko) [reply src/CSStrikes.sol:60]: I would prefer to keep it as is to be more solid and don't allow setting invalid data

File: src/interfaces/ICSModule.sol (hunk 7 above)
Comment (vgorkavenko) [reply src/interfaces/ICSModule.sol]: TODO added, because we don't know what values will be there yet

File: src/CSModule.sol [hunk 9]
```diff
@@ -862,6 +887,43 @@ contract CSModule is
...
+        }
+
+        uint256 pointer = _keyPointer(nodeOperatorId, keyIndex);
+        if (_isValidatorEjected[pointer]) revert AlreadySubmitted();
+
+        uint256 curveId = accounting.getBondCurveId(nodeOperatorId);
+
+        (, uint256 threshold) = PARAMETERS_REGISTRY.getStrikesParams(curveId);
+        bool isEjectable = _isEnoughStrikesToEject(strikesData, threshold);
+        if (!isEjectable) revert NotEnoughStrikesToEject();
```
[the remark this answers is not in the document]
Comment (vgorkavenko) [reply src/CSModule.sol]: Fixed

File: src/CSFeeOracle.sol [hunk 10]
```diff
@@ -38,14 +39,18 @@ contract CSFeeOracle is
     /// performance over the network computed by the off-chain oracle.
     uint256 internal _avgPerfLeewayBP;
 
+    ICSStrikes public strikes;
```
Comment (dgusakov) [inline src/CSFeeOracle.sol:42]: Re-use the slot above. I forgot to mention it as deprecated in my PR.

File: src/CSModule.sol [hunk 11]
```diff
@@ -85,6 +86,10 @@ contract CSModule is
     uint64 private _depositableValidatorsCount;
     uint64 private _nodeOperatorsCount;
 
+    ICSStrikes public strikes;
```
Comment (dgusakov) [inline src/CSModule.sol:89]: Maybe we can re-use the old deprecated slot with slashings?

Comment (dgusakov) [reply src/CSModule.sol:89]: In general I think we should find a way to set this address in one contract and then fetch by the others. Not sure how feasible it is, but let's try

File: src/CSModule.sol [hunk 12]
```diff
@@ -862,6 +887,43 @@ contract CSModule is
...
+        bytes32[] calldata proof
+    ) external {
+        _onlyExistingNodeOperator(nodeOperatorId);
+        NodeOperator storage no = _nodeOperators[nodeOperatorId];
+        if (keyIndex >= no.totalDepositedKeys) {
+            revert SigningKeysInvalidOffset();
+        }
+
+        uint256 pointer = _keyPointer(nodeOperatorId, keyIndex);
+        if (_isValidatorEjected[pointer]) revert AlreadySubmitted();
```
Comment (dgusakov) [inline src/CSModule.sol:904]: Maybe?
```suggestion
        if (_isValidatorEjected[pointer]) revert AlreadyEjected();
```

File: src/CSModule.sol [hunk 13]
```diff
@@ -862,6 +887,43 @@ contract CSModule is
...
+        uint256 penaltyAmount = PARAMETERS_REGISTRY.getBadPerformancePenalty(
+            curveId
+        );
+        if (penaltyAmount > 0)
+            accounting.penalize(nodeOperatorId, penaltyAmount);
+
+        _isValidatorEjected[pointer] = true;
+        emit EjectionSubmitted(nodeOperatorId, keyIndex, pubkey);
+
+        // TODO: call requestEjection(pointer);
```
Comment (dgusakov) [inline src/CSModule.sol:924]: Move upper. The event should go last

File: src/CSModule.sol (hunk 5 above)
Comment (dgusakov) [reply src/CSModule.sol]: Note about withdrawn is valid

File: src/CSModule.sol (hunk 12 above)
Comment (vgorkavenko) [reply src/CSModule.sol:904]: The current one is consistent with withdrawal submit, but we can change it

File: src/CSFeeOracle.sol (hunk 10 above)
Comment (vgorkavenko) [reply src/CSFeeOracle.sol:42]: We don't need to nullify it in `finalizeUpgradeV2` explicitly before assigning address, am I right?

File: src/CSModule.sol (hunk 11 above)
Comment (vgorkavenko) [reply src/CSModule.sol:89]: The only way I see is to change entry-point from `CSModule` to `CSStrikes`. It will help us to get rid of setting address in `CSModule` and just have one more role for calling `ejectBadPerformer`

Comment (vgorkavenko) [reply src/CSModule.sol:89]: Will implement it

File: src/CSFeeOracle.sol (hunk 10 above)
Comment (dgusakov) [reply src/CSFeeOracle.sol:35]: Yes

File: script/fork-helpers/SimulateVote.s.sol [hunk 14]
```diff
@@ -77,11 +77,16 @@ contract SimulateVote is Script, DeploymentFixtures, ForkHelpersCommon {
...
+
         OssifiableProxy csmProxy = OssifiableProxy(
             payable(deploymentConfig.csm)
         );
-        vm.broadcast(_prepareProxyAdmin(address(csmProxy)));
-        csmProxy.proxy__upgradeTo(upgradeConfig.csmImpl);
+        vm.startBroadcast(_prepareProxyAdmin(address(csmProxy)));
+        {
+            csmProxy.proxy__upgradeTo(upgradeConfig.csmImpl);
+            CSModule(deploymentConfig.csm).finalizeUpgradeV2();
```
Comment (dgusakov) [inline script/fork-helpers/SimulateVote.s.sol:87]: The EJECTION role is missing somewhere in the script

File: src/CSStrikes.sol [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        uint256[] calldata strikesData,
+        bytes32[] calldata proof
+    ) public view {
+        if (proof.length == 0) revert InvalidProof();
+        bool isValid = MerkleProof.verifyCalldata(
+            proof,
+            treeRoot,
+            hashLeaf(nodeOperatorId, pubkey, strikesData)
+        );
+        if (!isValid) revert InvalidProof();
```
Comment (skhomuti) [inline src/CSStrikes.sol:85]: it differs from the similar public view from `VettedGate`
i'd suggest making it returning boolean and check wherever it needed

File: src/CSStrikes.sol [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /// @inheritdoc ICSStrikes
+    function verifyProof(
+        uint256 nodeOperatorId,
+        bytes memory pubkey,
+        uint256[] calldata strikesData,
+        bytes32[] calldata proof
+    ) public view {
+        if (proof.length == 0) revert InvalidProof();
```
Comment (skhomuti) [inline src/CSStrikes.sol:79]: I believe you don't need to check it here, but dyor just in case

File: test/CSModule.t.sol [hunk 17]
```diff
@@ -5657,11 +5661,123 @@ contract CsmSubmitWithdrawal is CSMCommon {
...
+        uint256 keyIndex = 0;
+        uint256 noId = createNodeOperator();
+        (bytes memory pubkey, ) = csm.obtainDepositData(1, "");
+        csm.PARAMETERS_REGISTRY().setBadPerformancePenalty(0, 0);
+        uint256[] memory strikesData = new uint256[](6);
+        strikesData[0] = 1;
+        strikesData[1] = 2;
+        strikesData[5] = 3;
+
+        vm.expectEmit(true, true, true, true, address(csm));
```
Comment (skhomuti) [inline test/CSModule.t.sol:5706]: follow the trends
```suggestion
        vm.expectEmit(address(csm));
```

File: test/CSModule.t.sol [hunk 18]
```diff
@@ -5657,11 +5661,123 @@ contract CsmSubmitWithdrawal is CSMCommon {
...
+        emit ICSModule.EjectionSubmitted(noId, keyIndex, pubkey);
+        expectNoCall(
+            address(accounting),
+            abi.encodeWithSelector(accounting.penalize.selector, noId, 0)
+        );
+        csm.ejectBadPerformer(noId, keyIndex, strikesData.length);
+
+        bool ejected = csm.isValidatorEjected(noId, keyIndex);
+        assertTrue(ejected);
```
Comment (skhomuti) [inline test/CSModule.t.sol:5715]:
```suggestion
        assertTrue(csm.isValidatorEjected(noId, keyIndex));
```

File: test/CSModule.t.sol [hunk 19]
```diff
@@ -5657,11 +5661,123 @@ contract CsmSubmitWithdrawal is CSMCommon {
...
         csm.submitWithdrawal(noId, 0, depositSize, false);
-        vm.expectRevert(ICSModule.AlreadySubmitted.selector);
+        vm.expectRevert(ICSModule.AlreadyWithdrawn.selector);
         csm.submitWithdrawal(noId, 0, depositSize, false);
     }
 }
 
+contract CsmEjectBadPerformer is CSMCommon {
+    function test_ejectBadPerformer() public assertInvariants {
+        csm.grantRole(csm.BAD_PERFORMER_EJECTOR_ROLE(), address(this));
```
Comment (skhomuti) [inline test/CSModule.t.sol:5671]: grant it in the existing fixture

File: test/CSModule.t.sol [hunk 20]
```diff
@@ -5657,11 +5661,123 @@ contract CsmSubmitWithdrawal is CSMCommon {
...
+        csm.obtainDepositData(1, "");
+        uint256[] memory strikesData = new uint256[](2);
+        strikesData[0] = 1;
+        strikesData[1] = 2;
+
+        vm.expectRevert(ICSModule.NotEnoughStrikesToEject.selector);
+        csm.ejectBadPerformer(noId, 0, strikesData.length);
+    }
+
+    function test_ejectBadPerformer_RevertWhen_NoRole() public {
```
Comment (skhomuti) [inline test/CSModule.t.sol:5775]: move this test to `CSMAccessControl` contract

File: test/CSStrikes.t.sol [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+        (bytes memory pubkey, ) = keysSignatures(1);
+        uint256[] memory strikesData = new uint256[](6);
+        strikesData[0] = 100500;
+
+        tree.pushLeaf(abi.encode(noId, pubkey, strikesData));
+        tree.pushLeaf(abi.encode(noId + 1, pubkey, strikesData));
+
+        stdstore.target(address(strikes)).sig("treeRoot()").checked_write(
+            tree.root()
+        );
```
Comment (skhomuti) [inline test/CSStrikes.t.sol:97]: why not just set it using `processOracleReport`? it doesn't look complicated

File: test/CSStrikes.t.sol [hunk 22]
```diff
@@ -0,0 +1,279 @@
...
+
+        strikes.verifyProof({
+            nodeOperatorId: noId,
+            pubkey: pubkey,
+            strikesData: strikesData,
+            proof: proof
+        });
+    }
+
+    function test_verifyProof_RevertWhen_InconsistentData() public {
```
Comment (skhomuti) [inline test/CSStrikes.t.sol:109]: But I still think that it's better to return boolean here
```suggestion
    function test_verifyProof_RevertWhen_InvalidProof() public {
```

File: test/CSStrikes.t.sol [hunk 23]
```diff
@@ -0,0 +1,279 @@
...
+    function test_verifyProof_RevertWhen_EmptyProof() public {
+        uint256 noId = 42;
+        (bytes memory pubkey, ) = keysSignatures(1);
+        uint256[] memory strikesData = new uint256[](6);
+        strikesData[0] = 100500;
+
+        // Put a vulnerable `treeRoot` to make sure we provided the valid, but empty proof.
+        stdstore.target(address(strikes)).sig("treeRoot()").checked_write(
+            strikes.hashLeaf(noId, pubkey, strikesData)
+        );
```
Comment (skhomuti) [inline test/CSStrikes.t.sol:142]: I'm not sure if I get it
why it's vulnerable?

if you need to put any bytes here, it's better to put random bytes but not something that looks like a real value (tree leaf)

**Review by skhomuti — CHANGES_REQUESTED** [review]

File: src/CSModule.sol [hunk 24]
```diff
@@ -125,6 +130,9 @@ contract CSModule is
         _pauseFor(PausableUntil.PAUSE_INFINITELY);
     }
 
+    /// @dev should be called after update on the proxy
+    function finalizeUpgradeV2() external reinitializer(2) {}
```
Comment (dgusakov) [inline src/CSModule.sol:134]: Do we still need this method? It is now empty.

File: src/CSModule.sol [hunk 25]
```diff
@@ -106,11 +111,11 @@ contract CSModule is
         _disableInitializers();
     }
 
-    /// @notice initialize the module
+    /// @notice initialize the module from scratch
     function initialize(
         address _accounting,
         address admin
-    ) external initializer {
+    ) external reinitializer(2) {
```
Comment (dgusakov) [inline src/CSModule.sol:118]: I see no changes in the method body. Do we need to change the modifier here?

File: src/interfaces/ICSModule.sol [hunk 26]
```diff
@@ -479,6 +489,18 @@ interface ICSModule is IQueueLib, INOAddresses, IAssetRecovererLib {
         bool isSlashed
     ) external;
 
+    /// @notice Report Node Operator's key as withdrawn and settle withdrawn amount
```
Comment (dgusakov) [inline src/interfaces/ICSModule.sol:492]: This comment looks odd

File: test/CSParametersRegistry.t.sol [hunk 27]
```diff
@@ -1031,3 +1037,34 @@ contract CSParametersRegistryStrikesParamsTest is CSParametersRegistryBaseTest {
         assertEq(thresholdOut, defaultInitData.strikesThreshold);
     }
 }
+
+contract CSParametersRegistryBadPerformancePenaltyTest is
```
Comment (dgusakov) [inline test/CSParametersRegistry.t.sol:1041]: Pls, implement tests for the missing methods. Take the contracts above as an example

File: test/CSStrikes.t.sol [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+
+    function test_hashLeaf() public assertInvariants {
+        uint256 noId = 42;
+        (bytes memory pubkey, ) = keysSignatures(1);
+        uint256[] memory strikesData = new uint256[](6);
+        strikesData[0] = 100500;
+
+        assertEq(
+            strikes.hashLeaf(noId, pubkey, strikesData),
+            0x01eba5ed7fb9c5ebb4262844c7125628afcbea15e57bc7f4dd0c80d34d633584
```
Comment (dgusakov) [inline test/CSStrikes.t.sol:82]: Add a comment on how this const is calculated

File: test/CSStrikes.t.sol [hunk 29]
```diff
@@ -0,0 +1,279 @@
...
+        emit ICSStrikes.StrikesDataWiped();
+
+        vm.prank(oracle);
+        strikes.processOracleReport(bytes32(0), "");
+
+        assertEq(strikes.treeRoot(), bytes32(0));
+        assertEq(strikes.treeCid(), "");
+    }
+
+    function test_processOracleReport_NonEmptySubsequentReport()
```
Comment (dgusakov) [inline test/CSStrikes.t.sol:189]: Add a test when the first report is non empty and the next one is

File: test/fork/deployment/InitialState.t.sol [hunk 30]
```diff
@@ -74,6 +74,11 @@ contract ContractsInitialStateTest is Test, Utilities, DeploymentFixtures {
         );
     }
 
+    function test_strikes_initialState() public view {
+        assertEq(strikes.treeRoot(), bytes32(0));
+        assertEq(keccak256(abi.encodePacked(strikes.treeCid())), keccak256(""));
```
Comment (dgusakov) [inline test/fork/deployment/InitialState.t.sol:79]: Immutables asserts?

File: test/fork/deployment/PostDeployment.t.sol [hunk 31]
```diff
@@ -364,6 +369,29 @@ contract CSFeeDistributorDeploymentTest is Test, Utilities, DeploymentFixtures {
...
+    function setUp() public {
+        Env memory env = envVars();
+        vm.createSelectFork(env.RPC_URL);
+        initializeFromDeployment();
+        deployParams = parseDeployParams(env.DEPLOY_CONFIG);
+        adminsCount = block.chainid == 1 ? 1 : 2;
+    }
+
+    function test_constructor() public view {
+        assertEq(address(strikes.ORACLE()), address(oracle));
```
Comment (dgusakov) [inline test/fork/deployment/PostDeployment.t.sol:385]: Module?

File: test/helpers/Fixtures.sol [hunk 32]
```diff
@@ -108,7 +110,7 @@ contract DeploymentHelpers is Test {
         address permissionlessGate;
         address vettedGate;
         address parametersRegistry;
-        address parametersRegistryImpl;
+        address strikes;
```
Comment (dgusakov) [inline test/helpers/Fixtures.sol:113]: Why `parametersRegistryImpl` is removed?

File: test/CSStrikes.t.sol (hunk 23 above)
Comment (vgorkavenko) [reply test/CSStrikes.t.sol:142]: just copy-paste

File: src/CSModule.sol (hunk 24 above)
Comment (vgorkavenko) [reply src/CSModule.sol:134]: I think it is crucial to update contract version somehow

File: src/CSModule.sol (hunk 25 above)
Comment (vgorkavenko) [reply src/CSModule.sol:118]: If we want to have versioning we need to have right contract version after deploy from scratch

File: test/helpers/Fixtures.sol (hunk 32 above)
Comment (vgorkavenko) [reply test/helpers/Fixtures.sol:113]: couldn't find a place where it's used

File: test/fork/deployment/InitialState.t.sol (hunk 30 above)
Comment (vgorkavenko) [reply test/fork/deployment/InitialState.t.sol:79]: wdym?

File: test/CSStrikes.t.sol (hunk 21 above)
Comment (vgorkavenko) [reply test/CSStrikes.t.sol:97]: just copy-pasted thing

File: test/fork/deployment/InitialState.t.sol (hunk 30 above)
Comment (dgusakov) [reply test/fork/deployment/InitialState.t.sol:79]: Add asserts for the immutable parameters values

**Review by dgusakov — APPROVED** [review]

**Review by skhomuti — APPROVED** [review]
